l5kit/l5kit/evaluation/csv_utils.py

```diff
@@ -0,0 +1,246 @@
+    """
+    Generator function that returns a line at the time from the csv file as a dict
+
+    Args:
+        csv_path (str): path of the csv to read
+
+    Returns:
+        Iterator[dict]: dict keys are the csv header fieldnames
+    """
+    reader = csv.DictReader(open(csv_path, "r"))
+    fieldnames = reader.fieldnames
+    assert fieldnames is not None, "error reading fieldnames"
+
+    future_len = (len(fieldnames) - 2) / 3  # excluded first two fields, the rest must be (x, y, av) * len = 3*len
+    assert future_len == int(future_len), "error estimating len"
+    future_len = int(future_len)
+
+    coords_keys = _generate_coords_keys(future_len)
+    avails_keys = _generate_avails_keys(future_len)
+
+    for row in reader:
+        timestamp = row["timestamp"]
+        track_id = row["track_id"]
+
+        avail = np.asarray([np.float64(row[key]) for key in avails_keys])
+
+        coord = np.asarray([np.float64(row[key]) for key in coords_keys])
+        coord = coord.reshape((future_len, 2))
+
+        yield {"track_id": track_id, "timestamp": timestamp, "coord": coord, "avail": avail}
+
+
+def write_pred_csv(
+    csv_path: str,
+    timestamps: np.ndarray,
+    track_ids: np.ndarray,
+    coords: np.ndarray,
+    confs: Optional[np.ndarray] = None,
+) -> None:
+    """
+    Encode the ground truth into a csv file. Coords can have an additional axis for multi-mode.
+    We handle up to MAX_MODES modes.
+    For the uni-modal case, coords should not have the additional axis and confs should be set to None.
+    In this case, a single mode with confidence 1 will be written.
+
+    Args:
+        csv_path (str): path to the csv to write
+        timestamps (np.ndarray): (num_example,) frame timestamps
+        track_ids (np.ndarray): (num_example,) agent ids
+        coords (np.ndarray): (num_example x (modes) x future_len x num_coords) meters displacements
+        confs (Optional[np.ndarray]): (num_example x modes) confidence of each modes in each example.
+        Rows should sum to 1
+
+    Returns:
+
+    """
+    assert len(coords.shape) in [3, 4]
+
+    if len(coords.shape) == 3:
+        assert confs is None  # no conf for the single-mode case
+        coords = np.expand_dims(coords, 1)  # add a new axis for the multi-mode
+        confs = np.ones((len(coords), 1))  # full confidence
+
+    num_example, num_modes, future_len, num_coords = coords.shape
+    assert num_coords == 2
+    assert timestamps.shape == track_ids.shape == (num_example,)
+    assert confs is not None and confs.shape == (num_example, num_modes)
+    assert num_modes <= MAX_MODES
+
+    # generate always a fixed size json for MAX_MODES by padding the arrays with zeros
+    coords_padded = np.zeros((num_example, MAX_MODES, future_len, num_coords), dtype=coords.dtype)
+    coords_padded[:, :num_modes] = coords
+    confs_padded = np.zeros((num_example, MAX_MODES), dtype=confs.dtype)
+    confs_padded[:, :num_modes] = confs
+    # TODO check sum to 1
```

Comment:
Leave TODO in here?

Reply:
yes for now, I have a branch that fixes this (but remind it to me in a week xD)

Reply:
Do we need this check - which branch is it @lucabergamini ?
Should continue discussion there then, but my thoughts for now: What error tolerance do we allow?
Should our metric not penalize e.g. only confident predictions, and we leave it to the users how they would like to handle confidence?